lib/transfer-memory-allocator.js

```diff
@@ -0,0 +1,535 @@
+
+        this[INTERNAL].size = size;
+        this[INTERNAL].bufferPoolStartIndex = startIndex;
+
+        if (referencedMemoryFromBufferPool && size !== referencedMemoryFromBufferPool.length) {
+            throw new Error(`Requested buffer memory block size mismatch. Could not allocate Transfer Memory Block (requested: ${size}, allocated: ${referencedMemoryFromBufferPool.length})`);
+        }
+        this[INTERNAL].referencedMemoryFromBufferPool = referencedMemoryFromBufferPool;
+    }
+
+    /**
+     * Get the buffer size (size of allocated memory block)
+     * @returns {Integer} Size of the allocated buffer (memory block) 
+     */
+    get size() {
+        return this[INTERNAL].size;
+    }
+
+    /**
+     * Set the memory block size. Used by Transfer Memory Block allocator
+     * @param {Integer} value Size of allocated memory block
+     */
+    set size(value) {
+        this[INTERNAL].size = value;
+    }
+
+    /**
+     * Get the start index from the buffer pool (lowest index/memory
+     * location usable for this Transfer Memory Block)
+     * @returns {Integer} Start index of the allocated buffer 
+     * (memory block) in the buffer pool
+     */
+    get startIndex() {
+        return this[INTERNAL].bufferPoolStartIndex;
+    }
+
+    /**
+     * the start index from the buffer pool (lowest index/memory
+     * location usable for this Transfer Memory Block). 
+     * Used by Transfer Memory Block allocator.
+     * -1 means `unallocated` (or `deallocated`)
+     * @param {Integer} value Start index in buffer pool
+     */
+    set startIndex(value) {
+        this[INTERNAL].bufferPoolStartIndex = value;
+    }
+
+    /**
+     * Get the end index from the buffer pool (highest index/memory
+     * location usable for this Transfer Memory Block)
+     * @returns {Integer} End index of the allocated buffer 
+     * (memory block) in the buffer pool
+     */
+    get endIndex() {
+        return this.startIndex + this.size - 1;
+    }
+
+    /**
+     * Get the buffer representing the memory block from the pooled
+     * buffer allocated for this Transfer Memory Block.
+     * Modifying this memory block (buffer slice) will modify the memory 
+     * in the original Buffer because the allocated memory of this memory block 
+     * and the original buffer pool overlap.
+     */
+    get buffer() {
+        return this[INTERNAL].referencedMemoryFromBufferPool;
+    }
+
+    /**
+     * Set the buffer to use for memory access. 
+     * Used by Transfer Memory Block allocator
+     * @param {Buffer} value Buffer to use for memory access 
+     */
+    set buffer(value) {
+        return this[INTERNAL].referencedMemoryFromBufferPool = value;
+    }
+}
+
+// ------------------------------------------------------------------------------
+const PRIVATE = Symbol("private");
+/**
+ * Keeps track of memory usable for transfers, and allocates available 
+ * memory as needed/when requested by transfers
+ */
+class TransferMemoryBuffer {
```

Comment:
This is the class handling buffer memory allocations

Reply:
nth: split the classes into separate files

Reply:
No, because if I do that I need to make them all visible and exportable from the outside, which I don't want to. Especially `DeferredTransferMemoryBlockAllocation` should stay as a private class here. The only reason `TransferMemoryBlock` was also exported is to test a specific case (the segfault case - which happens when there is a mismatch between requested and allocated memory). But that will change as I'll rather use rewire for that specific test.